File: 2week/근우/homework/4.Minimum Difference Between Largest and Smallest Value in Three Moves.py
from typing import List


class Solution:
    def minDifference(self, nums: List[int]) -> int:
        if len(nums) < 4:
            return 0
        nums.sort()
        ans = 2e9
        for l in range(0, 4, 1):
            r = 3 - l
            ans = min(ans, nums[-1 - r] - nums[l])
        return ans


print(Solution().minDifference([1, 5, 0, 10, 14]))

File: 2week/남규/4. Minimum Difference Between Largest and Smallest Value in Three Moves.cpp
class Solution {
public:
    int minDifference(vector<int>& nums) {
        int n = nums.size();
        
        // TC1
        // [1,2,3]
        // return 0
        
        // TC2
        // [1,2,3,4,5]
        // return 1
        
        // TC3
        // [0, 3, 9, 10, 13, 18]
        // return 4
        
        // Time Complexity: O(N log N)
        // Space Complexity: O(1)
        
        if (n < 5) return 0;
        sort(nums.begin(), nums.end());
        int ans = INT_MAX;
        
        for(int i = 0; i < 4; i++){
            ans = min(ans, nums[n - 4 + i] - nums[i]);
        }
        return ans;
        
        // 0, 1, 5, 10, 14, 15
    }
};
